src/main/scala/com/fulcrumgenomics/vcf/AssessPhasing.scala

```diff
@@ -0,0 +1,780 @@
+      // NB: states is the reverse
+      states.reverse.toList
+    }
+
+    // backtrack
+    val states = {
+      if (scores.last._1 > scores.last._2) backtrack(0, from(scores.length-1)._1)
+      else if (scores.last._1 == scores.last._2) {
+        // choose the solution with the most # of long switch errors
+        val statesTop = backtrack(0, from(scores.length-1)._1)
+        val statesBot = backtrack(1, from(scores.length-1)._2)
+        val topScore  = numLongSwitchErrors(statesTop)
+        val botScore  = numLongSwitchErrors(statesBot)
+        if (topScore >= botScore) statesTop else statesBot
+      }
+      else backtrack(1, from(scores.length-1)._2)
+    }
+
+    //println("haplotypes: "  + states.toSeq.mkString(", "))
+    //println(s"states.length=${states.length} scores.length=${scores.length}")
+
+    // count how many mismatches are accounted for by point switch errors
+    val numPointErrors = {
+      require(phase.length == states.length)
+      phase.zip(states.map(_ == 0)).count { case (p, s) => p != s }
+    }
+
+    IlluminaSwitchErrors(numPointErrors, numLongSwitchErrors(states), states.length)
+  }
+
+  /** Partitions the cigar into multiple cigars, with each cigar being a contiguous block of phased variants. */
+  private[vcf] def toPhasedBlocks(cigar: Cigar, isTruth: Boolean = false): Seq[Cigar] = {
+    val endCigarTypes = if (isTruth) Set(TruthEnd, BothEnd) else Set(CallEnd, BothEnd)
+
+    // Splits the cigar any time we reach a CigarType in the endCigarTypes
+    val (cigarsToReturn, lastCigar) = cigar.foldLeft((ListBuffer[Cigar](), ListBuffer[CigarType]())) {
+      case ((cigars: ListBuffer[Cigar], currentCigar: ListBuffer[CigarType]), cigarType: CigarType) =>
+        cigarType match {
+          case tpe if endCigarTypes.contains(tpe) => // split!
+            if (currentCigar.nonEmpty) cigars.append(currentCigar.toList)
+            (cigars, new ListBuffer[CigarType]())
+          case _ => // keep going
+            currentCigar.append(cigarType)
+            (cigars, currentCigar)
+        }
+    }
+    // Make sure to ge the last one
+    if (lastCigar.nonEmpty) cigarsToReturn.append(lastCigar)
+
+    cigarsToReturn.toList
+  }
+
+  /** Returns the cigar for the two variant contexts. */
+  private[vcf] def contextsToCigar(t: Option[VariantContext], c: Option[VariantContext]): Cigar = (t, c) match {
+    case (None, Some(c: VCtx))          if isStartOfPhaseBlock(c)                             => Seq(CallEnd, Call)
+    case (None, Some(c: VCtx))                                                                => Seq(Call)
+    case (Some(t: VCtx), None)          if isStartOfPhaseBlock(t)                             => Seq(TruthEnd, Truth)
+    case (Some(t: VCtx), None)                                                                => Seq(Truth)
+    case (Some(t: VCtx), Some(c: VCtx)) if isStartOfPhaseBlock(t)  && isStartOfPhaseBlock(c)  => Seq(BothEnd, cigarTypeForVariantContexts(t, c))
```

Comment:
Doing these four cases as a single case that does another `match (isStart(t), isStart(c))` would prevent you from invoking those checks so frequently.